### PR TITLE
sealer join command

### PR DESCRIPTION
**Pull request by lllwan**

### CODE REVIEW

**Comment by cuisongliu** [comment]
the new file, pelease add lic , use 'make license' command generator it.

File: sealer/cmd/join.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	runArgs = &common.RunArgs{}
+	rootCmd.AddCommand(joinCmd)
+	joinCmd.Flags().StringVarP(&joinArgs.Masters, "masters", "m", "", "set Count or IPList to masters")
+	joinCmd.Flags().StringVarP(&joinArgs.Nodes, "nodes", "n", "", "set Count or IPList to nodes")
+	joinCmd.Flags().StringVarP(&joinArgs.User, "user", "u", "root", "set baremetal server username")
+	joinCmd.Flags().StringVarP(&joinArgs.Password, "passwd", "p", "", "set cloud provider or baremetal server password")
+	joinCmd.Flags().StringVarP(&joinArgs.Pk, "pk", "", cert.GetUserHomeDir()+"/.ssh/id_rsa", "set baremetal server private key")
+	joinCmd.Flags().StringVarP(&joinArgs.PkPassword, "pk-passwd", "", "", "set baremetal server  private key password")
+	joinCmd.Flags().StringVarP(&joinArgs.Interface, "interface", "i", "", "set default network interface name")
+	joinCmd.Flags().StringVarP(&clusterFile, "Clusterfile", "f", cert.GetUserHomeDir()+"/.sealer/my-cluster/Clusterfile", "apply a kubernetes cluster")
```
Comment (fanux) [inline sealer/cmd/join.go:65]: I think provide a -n(--cluster-name) flag is better.

File: sealer/cmd/join.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+		}
+	},
+}
+
+func init() {
+	runArgs = &common.RunArgs{}
+	rootCmd.AddCommand(joinCmd)
+	joinCmd.Flags().StringVarP(&joinArgs.Masters, "masters", "m", "", "set Count or IPList to masters")
+	joinCmd.Flags().StringVarP(&joinArgs.Nodes, "nodes", "n", "", "set Count or IPList to nodes")
+	joinCmd.Flags().StringVarP(&joinArgs.User, "user", "u", "root", "set baremetal server username")
```
Comment (fanux) [inline sealer/cmd/join.go:60]: User password pk pkpasswd and interface already in Clusterfile, should we provide those flags? I think if a user wants to update those configurations,just edit the file and using apply command.

File: sealer/cmd/join.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	sealer join --master 2 --node 3
+`,
+	Args: cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		if _, err := os.Lstat(clusterFile); err != nil {
+			logger.Error(clusterFile, err)
+			os.Exit(1)
+		}
+		applier := apply.JoinApplierFromArgs(clusterFile, joinArgs)
+		if applier == nil {
```
Comment (fanux) [inline sealer/cmd/join.go:45]: Maybe we can write an error log here

File: apply/join.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	cluster := &v1.Cluster{}
+	if err := utils.UnmarshalYamlFile(clusterfile, cluster); err != nil {
+		logger.Error("clusterfile parsing failed, please check:", err)
+		return nil
+	}
+	if joinArgs.Nodes == "" && joinArgs.Masters == "" {
+		logger.Error("The node or master parameter was not committed")
+		return nil
+	}
+	if cluster.Spec.Provider == "" {
```
Comment (fanux) [inline apply/join.go:46]: Provider enum:
ALI_CLOUD alibaba cloud
BAREMETAL exist host

File: sealer/cmd/join.go (hunk 3 above)
Comment (lllwan) [reply sealer/cmd/join.go:45]: It is not necessary because the superior function will print an error.

File: sealer/cmd/join.go (hunk 2 above)
Comment (lllwan) [reply sealer/cmd/join.go:60]: ok

File: sealer/cmd/join.go (hunk 1 above)
Comment (lllwan) [reply sealer/cmd/join.go:65]: ok

File: apply/join.go (hunk 4 above)
Comment (lllwan) [reply apply/join.go:46]: I'll use: baremetal

**Comment by fanux** [comment]
LGTM
/needs test

**Review by fanux — APPROVED** [review]

File: sealer/cmd/join.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+var joinCmd = &cobra.Command{
+	Use:   "join",
+	Short: "join node to cluster",
+	Example: `
+join to default cluster:
+	sealer join --master x.x.x.x --node x.x.x.x
+join to cluster by cloud provider, just set the number of masters or nodes:
+	sealer join --master 2 --node 3
+`,
+	Args: cobra.ExactArgs(1),
```
Comment (allencloud) [inline sealer/cmd/join.go:58]: I am afraid that `sealer join --master x.x.x.x --node x.x.x.x` needs no args.

File: sealer/cmd/join.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+)
+
+var clusterName string
+var joinArgs *common.RunArgs
+
+func getClusterName(sealerPath string) []string {
+	files, err := ioutil.ReadDir(sealerPath)
+	if err != nil {
+		logger.Error(err)
+		os.Exit(1)
```
Comment (allencloud) [inline sealer/cmd/join.go:38]: Please do not directly in an inner function. You could return an error, and make the function caller to decide which way to choose, exit or throw an error.

File: apply/join.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	if cluster.Spec.Provider == "BAREMETAL" {
+		if IsIPList(joinArgs.Nodes) || IsIPList(joinArgs.Masters) {
+			cluster.Spec.Masters.IPList = append(cluster.Spec.Masters.IPList, strings.Split(joinArgs.Masters, ",")...)
+			cluster.Spec.Nodes.IPList = append(cluster.Spec.Masters.IPList, strings.Split(joinArgs.Nodes, ",")...)
+		} else {
+			logger.Error("Parameter error:", "The current mode should submit iplist！")
+			return nil
+		}
+	} else {
```
Comment (allencloud) [inline apply/join.go:54]: } else if IsNumber(joinArgs.Nodes) || IsNumber(joinArgs.Masters) {
		cluster.Spec.Masters.Count = strconv.Itoa(StrToInt(cluster.Spec.Masters.Count) + StrToInt(joinArgs.Masters))
		cluster.Spec.Nodes.Count = strconv.Itoa(StrToInt(cluster.Spec.Nodes.Count) + StrToInt(joinArgs.Nodes))
	} else {
		logger.Error("Parameter error:", "The number of join masters or nodes that must be submitted to use cloud service！")
		return nil
	}

File: sealer/cmd/join.go (hunk 5 above)
Comment (lllwan) [reply sealer/cmd/join.go:58]: OK, I'll fix it

File: sealer/cmd/join.go (hunk 6 above)
Comment (lllwan) [reply sealer/cmd/join.go:38]: OK

File: apply/join.go (hunk 7 above)
Comment (lllwan) [reply apply/join.go:54]: OK

**Pull request merged**